core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -782,6 +782,12 @@ object IO {
         } yield fiberA.zipWith(fiberAs)(_ :: _)
     }
 
+  final def bracket[E, A, B](acquire: IO[E, A])(release: A => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
+    for {
+      a <- acquire.uninterruptibly
+      b <- use(a).ensuring(release(a).uninterruptibly)
+    } yield b
```

Comment:
This strategy will not work because the fiber can be interrupted between `a` and `b`, leading to acquisition but not release. In addition, the signature of this method should match the most powerful version of `bracket` available in `IO`, which is, I think, `bracket0`.

I would not refactor the tests at this point. This is a very ambitious ticket, I'm surprised anyone took it at all! 😉 

I was planning to take it myself. The best way to do this is incrementally, first by making a `bracket` that exactly re-creates the functionality of the existing `bracket0`, making sure it passes similar tests. Then the old one can be removed and replaced, which is something I'll need to help with because the primitive will become `Ensuring` not `Bracket`.

Reply:
> the fiber can be interrupted between `a` and `b`

Now I get why you had an `IORef[Option[ExitResult[E, A]]]` 👍 

> the signature of this method should match the most powerful version of `bracket` available in `IO`

Gotcha. Yeah I thought that `E` might be an ADT such that partially acquired resources could be released by a suitable process.

> I'm surprised anyone took it at all!

😅 I thought if I want to grok the code the only way is to tackle something ambitious, but I'll close this PR and keep working on my branch.